src/compiler/checker.ts

```diff
@@ -2616,20 +2593,18 @@ namespace ts {
                             return createTypeReferenceNode("Array", [typeArgumentNode]);
                         }
 
-                        context.flags |= NodeBuilderFlags.InElementType;
                         const elementType = typeToTypeNodeHelper(typeArguments[0], context);
-                        Debug.assert(!(context.flags & NodeBuilderFlags.InElementType));
-
                         return createArrayTypeNode(elementType);
                     }
                     else if (type.target.objectFlags & ObjectFlags.Tuple) {
                         if (typeArguments.length > 0) {
-                            const tupleConstituentNodes = mapToTypeNodeArray(typeArguments.slice(0, getTypeReferenceArity(type)), context, /*addInElementTypeFlag*/ false, /*addInFirstTypeArgumentFlag*/ false);
+                            const slice = typeArguments.slice(0, getTypeReferenceArity(type));

```

Comment:
If you add the check to `mapToTypeNodeArray` you don't need to break this up into two statements.